include/tsl/ordered_hash.h

```diff
@@ -1098,6 +1098,44 @@ class ordered_hash : private Hash, private KeyEqual {
     return 1;
   }
 
+  /**
+   * Clear a bucket without touching the container holding the values.
+   */
+  void clear_bucket(typename buckets_container_type::iterator it) {
+    tsl_oh_assert(it != m_buckets_data.end());
+    it->clear();
+    backward_shift(std::size_t(std::distance(m_buckets_data.begin(), it)));
+  }
+
+  /**
+   * Remove all entries for which the given predicate matches.
+   */
+  template <class Predicate>
+  size_type erase_if(Predicate &pred) {
+    using const_ref = typename values_container_type::const_reference;
+    auto last = m_values.end();
+    auto first = m_values.begin();
+    for (; first != last; ++first) {
```

Comment:
I would prefer to avoid the double-loop and use something like:
```c++
    auto first = std::find_if(m_values.begin(), m_values.end(), pred);
    if (first == m_values.end()) {
      return 0;
    }

    clear_bucket(find_key(KeySelect()(*first), hash_key(KeySelect()(*first))));
    for (auto it = first + 1; it != m_values.end(); ++it) {
```

with an early exit.

Reply:
I rewrote the loop like this because of the constness issue below.

Reply:
This also means one unnecessary comparison but the compiler should be able to optimize it away.

Reply:
Why is the constness issue below problematic? A mutable reference will not work anyway due to the later `pred(static_cast<const_ref>(*it))` call.

Reply:
> Why is the constness issue below problematic? A mutable reference will not work anyway due to the later `pred(static_cast<const_ref>(*it))` call.

On could still overload the lambda, right?

Reply:
We could of course also put another lambda in between:

```c++
auto cpred = [&pred](const_ref x) { return pred(x); };
```

Then the function could be refactored as you suggest.

Reply:
> On could still overload the lambda, right?

Oh yeah you're right.

Another way would be to use `std::find(m_values.cbegin(), m_values.cend(), pred);` and convert the resulting iterator with `mutable_iterator`.

Reply:
I don't think there is a nice way to convert a const_iterator into an iterator. (I can only think of `begin() + distance(...)`). It only works the other way round.

Reply:
There's a `mutable_iterator` method in `ordered_hash.h` that does it like that.

But otherwise your solution with a second lambda or even just have a normal find loop instead of using `find_if` would be good.

Reply:
I prefer the lamba or loop over the mutable_iterator because there is no overhead involved. I don't think the compiler can optimize the mutable_iterator call away.